### ISSUE TITLE
CSS custom properties issue

### ISSUE THREAD

**Issue by zce** [system-ui/theme-ui#499]
```css
body {
  --theme-ui-colors-text: var(--theme-ui-colors-text, #495057);
  /* --theme-ui-colors-text => undefined */
}
```

Maybe the right way

```css
body {
  --theme-ui-colors-text: #495057;
  /* --theme-ui-colors-text => #495057 */
}
```

https://codepen.io/zce/pen/rNaaByE

**Comment by zce** [comment]
I will try to submit a PR

**Comment by jxnblk** [comment]
I'm not sure I understand what the issue here is. Could you provide a reproduction?

**Comment by zce** [comment]
I'm sorry, I may not have made it clear.

I mean, there is something wrong with the current color styles.

The color style currently generated is like this

```css
body {
  --theme-ui-colors-text: var(--theme-ui-colors-text, #495057);
  /* --theme-ui-colors-text => undefined */
}
```

In this case `--theme-ui-colors-text` will get `undefined`, then `--theme-ui-colors-text` loses meaning.

At present, all sites that use theme-ui are all in this situation.

![image](https://user-images.githubusercontent.com/6166576/70103774-f88a8100-1676-11ea-9649-a35d5610474a.png)

**Comment by jxnblk** [comment]
I'm still not sure what you're describing. Can you provide a reproduction in Code Sandbox or another repo?

**Comment by zce** [comment]
https://codesandbox.io/s/infallible-hawking-sotbj

https://github.com/zce/custom-properties-issue

**Comment by jxnblk** [comment]
Thanks! First thing I'm noticing is that you *shouldn't* be using the internal custom properties directly, e.g. `color: "var(--theme-ui-colors-primary)"` should be `color: 'primary'`

**Comment by zce** [comment]
I know the right usage, I want to show that I can't get the custom properties defined in body at all.

Because css variables cannot be defined correctly in following way,

```css
body {
  --theme-ui-colors-text: var(--theme-ui-colors-text, #495057);
}
```

The right way is

```css
body {
  --theme-ui-colors-text: #495057;
}
```

**Comment by jxnblk** [comment]
I don’t understand where you’re seeing that first example occur with theme-ui. Does your example produce that?

**Comment by zce** [comment]
All sites using theme-ui in production mode, The global style generated by `<ColorMode/>` component.

![image](https://user-images.githubusercontent.com/6166576/70285640-256b9f00-1803-11ea-8759-0a58e811a360.png)

Development mode

![image](https://user-images.githubusercontent.com/6166576/70285843-de31de00-1803-11ea-99ef-98eeb9d16b28.png)

The reason for this problem is that [toCustomProperties](https://github.com/system-ui/theme-ui/blob/cf0b9c4b51e14665280d062f53c7bd50e0e667b1/packages/theme-ui/src/custom-properties.js#L26-L45) execute before [createColorStyles](https://github.com/system-ui/theme-ui/blob/cf0b9c4b51e14665280d062f53c7bd50e0e667b1/packages/theme-ui/src/custom-properties.js#L26-L45)

**Comment by zce** [comment]
At present, the default value of var function is working in all places where color is used.

**Comment by zce** [comment]
This is the cause you want to use custom properties in style files in third-party libraries

**Comment by jxnblk** [comment]
The custom properties in this library are *not* intended to be used outside this library, if that's what you're trying to do. If you can create a failing test for what you're trying to describe, maybe that would help

**Comment by zce** [comment]
Thanks, What is the original intention of custom properties?

Only for color flash?

This is how I use it now

https://github.com/zce/zce.me/blob/fcd9af175b0cf293fda824aa4418a5889ea43792/src/utils/styles.js#L21

**Comment by zce** [comment]
I thought it was an oversight because what I saw in the test cases was this

https://github.com/system-ui/theme-ui/blob/cf0b9c4b51e14665280d062f53c7bd50e0e667b1/packages/theme-ui/test/custom-properties.js#L61

**Comment by jxnblk** [comment]
Closing this because the internal custom properties implementation will be changing slightly in v0.3

**Comment by iampava** [comment]
Hey!

If I understand correctly, it's intended for the CSS variables to not be used directly by us. However, shouldn't we allow this option (converting all theme props to CSS Variables and inserting in page)? 

I arrived here after trying to style common elements like `<h1>`, `<h2>`, `<strong>`, `<a>` according to the theme and then just use them directly in the page. Later, when a 3rd party app embeds our code, it will change some part of the theme, but since it's CSS Variables these elements will update as well.

I hacked together this adapter which allow dynamic insertion of CSS Variables.  (just colors for now)

```jsx
export const ThemeAdapter = (props) => {
  const ref = useRef(null);

  useEffect(() => {
    const { theme } = props;

    if (theme && theme.colors) {
      Object.keys(theme.colors).forEach(key => {
        ref.current.style.setProperty(`--colors-${key}`, theme.colors[key]);
      })
    }
  });

  return (
    <div ref={ref} >
      <ThemeProvider {...props} />
    </div>
  )
}
```

Although this wasn't hard, I wonder whether this option should be present by default.
